**try语句块和异常处理**

​		异常处理机制为程序中异常检测和异常处理这两部分的协作提供支持。

- throw表达式

  异常检测部分使用throw表达式来表示它遇到了无法处理的问题。

- try语句块

  异常处理部分使用try语句块处理异常。try语句块以try关键字开始，并以一个或多个catch子句结束。

1、throw表达式

​		throw表达式包含关键字throw和紧随其后的一个表达式。，其中表达式的类型就是抛出的异常的类型。

```C++
throw runtime_error("XXXXX");//抛出一个runtime_error类型的异常
```

​		runtime_error是标准库异常类型的一种，定义在**stdexcept**头文件中。

2、try语句块

try{

}catch{

}catch{

}……

**标准异常**

- exception头文件中定义了最通用的异常类exception。他只报告异常的发生，不提供任何额外的信息。
- stdexcept头文件定义了几种常用的异常类
- new头文件定义了bad_alloc异常类型
- type_info头文件定义了bad_cast异常类型

​		标准库异常类只定义了几种运算，包括**创建或拷贝**异常类型的对象，以及为对象的**赋值**。

​		只能以默认初始化的方式初始化exception、bad_alloc、和bad_cast对象，不允许为这些对象提供初始值。

​		其他异常类则恰恰相反：应该使用string对象或者C风格的字符串初始化这些类型的对象。不允许使用默认初始化的方式。

​		out_of_range定义于<stdexcept>

​		异常类型只定义了一个名为what的成员函数，该函数没有任何的参数，返回值是一个指向C风格字符串的const char * 。 what函数返回的C风格字符串的内容与异常对象的类型有关。如果异常类型有一个字符串初始值，则what返回该字符串。对于其他无初始值的异常类型，what的返回内容由编译器决定。

